pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+        arguments.extend([inspector.add_sets(insp_sets)])
+        arguments.extend([inspector.add_maps(insp_maps.values())])
+        inspector.add_loops(insp_loops)
+        # Get type and value of any additional arguments that the SLOPE's inspector
+        # expects
+        arguments.extend([inspector.set_external_dats()])
+
+        # Set a specific tile size
+        arguments.extend([inspector.set_tile_size(self._tile_size)])
+
+        # Arguments types and values
+        argtypes, argvalues = zip(*arguments)
+
+        # Generate inspector C code
+        src = inspector.generate_code()
+
+        # Return type of the inspector
+        rettype = slope.Executor.meta['py_ctype_exec']
+
+        # Compiler and linker options
+        slope_dir = os.environ['SLOPE_DIR']
+        compiler = coffee.plan.compiler.get('name')
+        cppargs = slope.get_compile_opts(compiler)
+        cppargs += ['-I%s/%s' % (slope_dir, slope.get_include_dir())]
+        ldargs = ['-L%s/%s' % (slope_dir, slope.get_lib_dir()),
+                  '-l%s' % slope.get_lib_name()]
+
+        # Compile and run inspector
+        fun = compilation.load(src, "cpp", "inspector", cppargs, ldargs,
+                               argtypes, rettype, compiler)
+        inspection = fun(*argvalues)
+
+        # Finally, get the Executor representation, to be used at executor's
+        # code generation time
+        executor = slope.Executor(inspector)
+
+        self._schedule = TilingSchedule(self._schedule, inspection, executor)
+
+    @property
+    def mode(self):
+        return self._mode
+
+
+# Interface for triggering loop fusion
+
+def fuse(name, loop_chain, tile_size):
+    """Apply fusion (and possibly tiling) to a list of :class:`ParLoop` obecjts,
+    which we refer to as ``loop_chain``. Return a smaller list of :class:`ParLoop`s
+    objects, when some loops may have been fused/tiled. If fusion could not be
+    applied, return the original, unmodified ``loop_chain``.
+
+    .. note::
+       At the moment, the following features are not supported, in which
+       case the unmodified ``loop_chain`` is returned.
+
+        * mixed ``Datasets`` and ``Maps``;
+        * extruded ``Sets``
+
+    .. note::
+       Tiling cannot be applied if any of the following conditions verifies:
+
+        * a global reduction/write occurs in ``loop_chain``
+    """
+    # If there is nothing to fuse, just return
+    if len(loop_chain) in [0, 1]:
+        return loop_chain
+
+    # Search for _Assembly objects since they introduce a synchronization point;
+    # that is, loops cannot be fused across an _Assembly object. In that case, try
+    # to fuse only the segment of loop chain right before the synchronization point
+    remainder = []
+    synch_points = [l for l in loop_chain if isinstance(l, Mat._Assembly)]
+    if synch_points:
+        if len(synch_points) > 1:
+            warning("Fusing loops and found more than one synchronization point")
```

Comment:
Yes, it should be correct, since you're not throwing away anything. It's just that we don't try to fuse after the second synch point. I think this is plausible, because in practice it should never occur